Sources/GRPCCore/Call/ClientResponse.swift

```diff
@@ -0,0 +1,436 @@
+  ///
+  /// For rejected RPCs (in other words, where ``accepted`` is `failure`) the metadata is empty.
+  public var metadata: Metadata {
+    switch self.accepted {
+    case let .success(contents):
+      return contents.metadata
+    case .failure:
+      return [:]
+    }
+  }
+
+  /// Returns the message received from the server.
+  ///
+  /// - Throws: ``RPCError`` if the request failed.
+  public var message: Message {
+    get throws {
+      try self.accepted.map { $0.message }.get()
+    }
+  }
+
+  /// Returns metadata received from the server at the end of the response.
+  ///
+  /// Unlike ``metadata``, for rejected RPCs the metadata returned may contain values.
+  public var trailingMetadata: Metadata {
+    switch self.accepted {
+    case let .success(contents):
+      return contents.trailingMetadata
+    case let .failure(error):
+      return error.metadata
+    }
+  }
+}
+
+@available(macOS 10.15, iOS 13, tvOS 13, watchOS 6, *)
+extension ClientResponse.Stream {
+  /// Creates a new accepted response.
+  ///
+  /// - Parameters:
+  ///   - messageType: The type of message.
+  ///   - metadata: Metadata received from the server at the beginning of the response.
+  ///   - bodyParts: An ``RPCAsyncSequence`` of response parts received from the server.
+  public init(
+    of messageType: Message.Type = Message.self,
+    metadata: Metadata,
+    bodyParts: RPCAsyncSequence<Contents.BodyPart>
+  ) {
+    let contents = Contents(metadata: metadata, bodyParts: bodyParts)
+    self.accepted = .success(contents)
+  }
+
+  /// Creates a new failed response.
+  ///
+  /// - Parameters:
+  ///   - messageType: The type of message.
+  ///   - error: An error describing why the RPC failed.
+  public init(of messageType: Message.Type = Message.self, error: RPCError) {
+    self.accepted = .failure(error)
+  }
+
+  /// Returns metadata received from the server at the start of the response.
+  ///
+  /// For rejected RPCs (in other words, where ``accepted`` is `failure`) the metadata is empty.
+  public var metadata: Metadata {
+    switch self.accepted {
+    case let .success(contents):
+      return contents.metadata
+    case .failure:
+      return [:]
+    }
+  }
+
+  /// Returns metadata received from the server at the end of the response.
+  ///
+  /// Unlike ``metadata``, for rejected RPCs the metadata returned may contain values.
+  public var messages: RPCAsyncSequence<Message> {
+    switch self.accepted {
+    case let .success(contents):
+      let filtered = contents.bodyParts.compactMap {
+        switch $0 {
+        case let .message(message):
+          return message
+        case .trailingMetadata:
+          return nil
+        }
+      }
+
+      return RPCAsyncSequence(wrapping: filtered)
+
+    case let .failure(error):
+      return RPCAsyncSequence.throwing(error)
+    }
+  }
+}
+
+// MARK: - Conversion
+
+@available(macOS 10.15, iOS 13, tvOS 13, watchOS 6, *)
+extension ClientResponse.Single {
+  @_spi(Testing)
```

Comment:
In the future we should probably switch to `package` access for those, but that is gated behind Swift versions.

Reply:
Agreed. I think it's best to use `_spi` for now since it isn't version specific. We can revisit this when we get close the releasing or when 5.9 is our minimum supported version.